### PR TITLE
Merged archive extraction routines.

### LINKED ISSUE #420: NuGet do not honor time-stamps when unpacking nupkg files
NuGet uses the current time when unpacking nupkg files, instead of time-stamps stored in the ZIP files.

This leads to "dirty file" problems for build & deployment systems relying on time-stamps for figuring out which header/lib/dll files have changed. An example of this is robocopy-based SW deployment that only replaces new/modified files.

Steps to reproduce:
* With NuGet command-line: "NuGet.exe install zlib" (package found on nuget.org), or
* Create a Visual Studio 2013 C++ project. Add NuGet dependency to "zlib" and build project.
* Verify that unpacked *.h, *.lib & *.dll files are time-stamped with current time instead of the time in the downloaded nupkg files.

### PR DESCRIPTION
**Pull request by alpaix**
Fixes https://github.com/NuGet/Home/issues/420.
- Refactored two file extraction methods into one.
- Added nuget.exe integration test to verify the fix.

### CODE REVIEW

File: src/NuGet.Core/NuGet.Packaging/PackageExtraction/NuGetPackageUtils.cs [hunk 1]
```diff
@@ -282,29 +282,7 @@ private static bool ShouldInclude(string fullName, string nupkgFileName, string
...
-                        using (var targetStream
-                            = new FileStream(targetFile, FileMode.Create, FileAccess.Write, FileShare.None))
-                        {
-                            entryStream.CopyTo(targetStream);
-                        }
-                    }
-
-                    File.SetLastWriteTimeUtc(targetFile, entry.LastWriteTime.UtcDateTime);
-                }
+                await PackageHelper.CreatePackageFileAsync(targetFile, entry, CancellationToken.None);
```
Comment (pranavkm) [inline src/NuGet.Core/NuGet.Packaging/PackageExtraction/NuGetPackageUtils.cs:285]: You could declare this method to not be async and return the result of `PackageHelper.CreatePackageFileAsync(targetFile, entry, CancellationToken.None);`

File: src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageHelper.cs [hunk 2]
```diff
@@ -237,26 +243,23 @@ public static IEnumerable<ZipFilePair> GetInstalledPackageFiles(IEnumerable<ZipF
                 return packageFileFullPath;
             }
 
-            using (Stream outputStream = File.Create(packageFileFullPath))
+            using (Stream outputStream = new FileStream(packageFileFullPath, FileMode.Create, FileAccess.Write, FileShare.None))
```
Comment (pranavkm) [inline src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageHelper.cs]: Need to specify `useAsync: true`

Comment (alpaix) [inline src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageHelper.cs]: Sure. Will add it.

File: src/NuGet.Core/NuGet.Packaging/PackageExtraction/NuGetPackageUtils.cs (hunk 1 above)
Comment (alpaix) [inline src/NuGet.Core/NuGet.Packaging/PackageExtraction/NuGetPackageUtils.cs:285]: I prefer having explicit await statement and let the compiler optimize it.

File: src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageHelper.cs [hunk 3]
```diff
@@ -142,39 +142,28 @@ public static bool IsSatellitePackage(NuspecReader nuspecReader, out PackageIden
             return false;
         }
 
-        public static Task<IEnumerable<ZipFilePair>> GetPackageFiles(IEnumerable<ZipArchiveEntry> packageFiles, string packageDirectory,
-            PackageSaveModes packageSaveMode, CancellationToken token)
+        public static IEnumerable<ZipFilePair> EnumeratePackageFiles(IEnumerable<ZipArchiveEntry> packageEntries, string packageDirectory,
+            PackageSaveModes packageSaveMode)
```
Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageHelper.cs:146]: looks like packageSaveMode is not used either?

File: src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageHelper.cs [hunk 4]
```diff
@@ -186,49 +175,66 @@ public static IEnumerable<ZipFilePair> GetInstalledPackageFiles(IEnumerable<ZipF
...
+                var packageFiles = EnumeratePackageFiles(zipArchive.Entries, packageDirectory, packageSaveMode);
+                installedPackageFiles = GetInstalledPackageFiles(packageFiles);
             }
 
-            return installedPackageFiles;
+            return installedPackageFiles.ToList();
         }
 
-        public static async Task<Tuple<string, IEnumerable<ZipFilePair>>> GetInstalledSatelliteFiles(Stream packageStream,
+        public static Tuple<string, IEnumerable<ZipFilePair>> GetInstalledSatelliteFiles(Stream packageStream,
```
Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageHelper.cs:195]: while you are at it, get rid of the tuple, also I don't see packageSaveMode being used.

File: src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageHelper.cs (hunk 2 above)
Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageHelper.cs]: var

File: src/NuGet.Core/NuGet.ProjectManagement/Projects/FolderNuGetProject.cs [hunk 5]
```diff
@@ -164,7 +164,7 @@ public string GetInstalledPath(PackageIdentity packageIdentity)
             return PackagePathResolver.GetInstalledPath(packageIdentity) ?? string.Empty;
         }
 
-        public async Task<bool> DeletePackage(PackageIdentity packageIdentity,
+        public Task<bool> DeletePackage(PackageIdentity packageIdentity,
```
Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.ProjectManagement/Projects/FolderNuGetProject.cs:167]: keep the method async (so the caller will get an error if he doesn't await) and then exceptions are going to be properly unrwapped

File: src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageHelper.cs (hunk 3 above)
Comment (alpaix) [inline src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageHelper.cs:146]: actually it is used (see line 152) in IsPackageFile call.

**Comment by yishaigalatzer** [comment]
:shipit:

File: src/NuGet.Core/NuGet.ProjectManagement/Projects/FolderNuGetProject.cs (hunk 5 above)
Comment (alpaix) [inline src/NuGet.Core/NuGet.ProjectManagement/Projects/FolderNuGetProject.cs:167]: cannot make the method async as it lacks awaits

Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.ProjectManagement/Projects/FolderNuGetProject.cs:167]: kk

File: src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageHelper.cs (hunk 4 above)
Comment (alpaix) [inline src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageHelper.cs:195]: it is whole lot of refactoring as it's deeply integrated with several methods in a call chain.
I'd prefer to avoid doing it unless it is really needed.

Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageHelper.cs:195]: I agree if it is large, it should be done in a separate PR. Should be two easy steps
1. Create a class with the right type that has Item1 & Item2 on it.
2. Then use autorefactoring when you rename Item1 -> something, Item2 -> something else

Comment (alpaix) [inline src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageHelper.cs:195]: Got it. I misinterpreted you comment. I thought you wanted me to omit one of tuple parameters as if it was not used along the way.

**Comment by forderud** [comment]
@alpaix I've now tried to build this PR, and install the generated NuGet.Tools.vsix. Unfortunately, the reproducer uploaded to https://github.com/NuGet/Home/issues/420#issuecomment-167266769 still reproduces the issue.

Are you able to verify that this PR fixes my reproducer?

**Comment by alpaix** [comment]
@forderud I tested the fix with nuget.exe restore. I was able to verify it
image

Will attempt to install VS extension and repro the issue as well.

**Comment by forderud** [comment]
Strange... I still get the time-stamp issue even with "nuget.exe restore" from the command line. If the problem is fixed for you, then I'm probably experiencing a local problem related to my NuGet cache and/or build environment.

Please move ahead and merge the pull request if it works for you. I can verify the fix afterwards if there is some kind of nightly build of the NuGet "dev" branch available for download?

**Comment by yishaigalatzer** [comment]
Yes - go to myget.org -> nugetbuild feed -> NuGet.CommandLine package
